src/operations/SymbolicOperation.cpp

```diff
@@ -0,0 +1,286 @@
+                                      [[maybe_unused]] const Symbolic& theta) {
+        // parse a real u3 gate
+        checkInteger(phi);
+        checkFractionPi(phi);
+
+        return U3;
+    }
+    OpType SymbolicOperation::parseU3(dd::fp& lambda, [[maybe_unused]] const Symbolic& phi,
+                                      [[maybe_unused]] const Symbolic& theta) {
+        // parse a real u3 gate
+        checkInteger(lambda);
+        checkFractionPi(lambda);
+
+        return U3;
+    }
+
+    OpType SymbolicOperation::parseU2([[maybe_unused]] const Symbolic& lambda, [[maybe_unused]] const Symbolic& phi) {
+        return U2;
+    }
+
+    OpType SymbolicOperation::parseU2([[maybe_unused]] const Symbolic& lambda, dd::fp& phi) {
+        checkInteger(phi);
+        checkFractionPi(phi);
+
+        return U2;
+    }
+    OpType SymbolicOperation::parseU2(dd::fp& lambda, [[maybe_unused]] const Symbolic& phi) {
+        checkInteger(lambda);
+        checkFractionPi(lambda);
+
+        return U2;
+    }
+
+    OpType SymbolicOperation::parseU1([[maybe_unused]] const Symbolic& lambda) {
+        return Phase;
+    }
+
+    void SymbolicOperation::checkUgate() {
+        if (type == Phase) {
+            if (!isSymbolicParameter(0))
+                type = StandardOperation::parseU1(parameter[0]);
+        } else if (type == U2) {
+            if (!isSymbolicParameter(0) && !isSymbolicParameter(1))
+                type = StandardOperation::parseU2(parameter[0], parameter[1]);
+            else if (isSymbolicParameter(0))
+                type = parseU2(symbolicParameter[0].value(), parameter[1]);
+            else if (isSymbolicParameter(1))
+                type = parseU2(parameter[1], symbolicParameter[1].value());
+        } else if (type == U3) {
+            if (!isSymbolicParameter(0) && !isSymbolicParameter(1) && !isSymbolicParameter(2))
+                type = StandardOperation::parseU3(parameter[0], parameter[1], parameter[2]);
+            else if (!isSymbolicParameter(0) && !isSymbolicParameter(1))
+                type = parseU3(parameter[0], parameter[1], symbolicParameter[2].value());
+            else if (!isSymbolicParameter(0) && !isSymbolicParameter(2))
+                type = parseU3(parameter[0], symbolicParameter[1].value(), parameter[2]);
+            else if (!isSymbolicParameter(1) && !isSymbolicParameter(2))
+                type = parseU3(symbolicParameter[0].value(), parameter[1], parameter[2]);
+            else if (!isSymbolicParameter(0))
+                type = parseU3(parameter[0], symbolicParameter[1].value(), symbolicParameter[2].value());
+            else if (!isSymbolicParameter(1))
+                type = parseU3(symbolicParameter[0].value(), parameter[1], symbolicParameter[2].value());
+            else if (!isSymbolicParameter(2))
+                type = parseU3(symbolicParameter[0].value(), symbolicParameter[1].value(), parameter[2]);
+        }
```

Comment:
this does look a little messy to be honest. this also holds for all the `parseU...` functions with different signatures.
Wouldn't it be cleaner to have individual `parseU...` functions that all take `const SymbolOrNumber&`s and then correspondingly treat the different cases?
In my imagination that would safe some of the code duplication above.

Reply:
But wouldn't that just offset the problem to the next function call? Due to the different nature of symbols and numbers I always have to figure out which combination of symbols and numbers I have right now.

Reply:
Technically that is true. I still somehow feel that it is more straight-forward to follow the same structure as in the base class function and essentially break out of checks whenever a parameter is symbolic and, hence, the simplification condition cannot be decided.
In the end, I am fine with the current state. So if you are not convinced by my argumentation above, then feel free to leave this as it is.